File: frontend-obrasmart2/src/app/service/distancia.service.ts
// distancia.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';

export interface DistanciaCalculo {
  equipoId: number;
  equipoNombre: string;
  equipoLatitud: number;
  equipoLongitud: number;
  obradorId: number;
  obradorNombre: string;
  obradorLatitud: number;
  obradorLongitud: number;
  distanciaKm: number;
  mensaje: string;
}

@Injectable({
  providedIn: 'root'
})
export class DistanciaService {
  
  private apiUrl = `${environment.apiUrl}/distancias`;

  constructor(private http: HttpClient) { }

  /**
   * Calcula la distancia entre un equipo y un obrador
   * @param equipoId ID del equipo
   * @param obradorId ID del obrador
   * @returns Observable con la información de distancia
   */
  calcularDistancia(equipoId: number, obradorId: number): Observable<DistanciaCalculo> {
    const params = new HttpParams()
      .set('equipoId', equipoId.toString())
      .set('obradorId', obradorId.toString());
    
    return this.http.get<DistanciaCalculo>(`${this.apiUrl}/calcular`, { params });
  }

  /**
   * Calcula distancia usando la fórmula de Haversine directamente en el frontend
   * Útil para cálculos rápidos sin llamar al backend
   */
  calcularDistanciaLocal(
    lat1: number, lon1: number, 
    lat2: number, lon2: number
  ): number {
    const R = 6371; // Radio de la Tierra en km
    const dLat = this.toRad(lat2 - lat1);
    const dLon = this.toRad(lon2 - lon1);
    
    const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
              Math.cos(this.toRad(lat1)) * Math.cos(this.toRad(lat2)) *
              Math.sin(dLon / 2) * Math.sin(dLon / 2);
    
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    const distancia = R * c;
    
    return Math.round(distancia * 100) / 100; // Redondear a 2 decimales
  }

  private toRad(grados: number): number {
    return grados * (Math.PI / 180);
  }
}

File: GestionLogistica/src/main/java/com/ObraSmart/GestionLogistica/service/impl/DistanciaService.java
package com.ObraSmart.GestionLogistica.service.impl;


import com.ObraSmart.GestionLogistica.client.EquipoClient;
import com.ObraSmart.GestionLogistica.client.ObradorClient;
import com.ObraSmart.GestionLogistica.dto.DistanciaCalculoDto;
import com.ObraSmart.GestionLogistica.dto.EquipoSimpleDto;
import com.ObraSmart.GestionLogistica.dto.ObradorSimpleDto;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
@Slf4j
public class DistanciaService {

    private final EquipoClient equipoClient;
    private final ObradorClient obradorClient;

    // Radio de la Tierra en kilómetros
    private static final double RADIO_TIERRA_KM = 6371.0;

    /**
     * Calcula la distancia entre un equipo y un obrador usando la fórmula de Haversine
     *
     * @param equipoId ID del equipo
     * @param obradorId ID del obrador
     * @return DTO con información de distancia y coordenadas
     */
    public DistanciaCalculoDto calcularDistancia(Long equipoId, Long obradorId) {
        log.info("Calculando distancia entre equipo {} y obrador {}", equipoId, obradorId);

        // Obtener información del equipo
        EquipoSimpleDto equipo = equipoClient.obtenerEquipo(equipoId);

        // Obtener información del obrador
        ObradorSimpleDto obrador = obradorClient.obtenerObrador(obradorId);

        // Calcular distancia usando Haversine
        double distanciaKm = calcularDistanciaHaversine(
                equipo.getLatitud(),
                equipo.getLongitud(),
                obrador.getLatitud(),
                obrador.getLongitud()
        );

        // Construir respuesta
        DistanciaCalculoDto resultado = DistanciaCalculoDto.builder()
                .equipoId(equipo.getId())
                .equipoNombre(equipo.getNombre())
                .equipoLatitud(equipo.getLatitud())
                .equipoLongitud(equipo.getLongitud())
                .obradorId(obrador.getId())
                .obradorNombre(obrador.getNombre())
                .obradorLatitud(obrador.getLatitud())
                .obradorLongitud(obrador.getLongitud())
                .distanciaKm(Math.round(distanciaKm * 100.0) / 100.0) // Redondear a 2 decimales
                .mensaje(generarMensaje(equipo.getNombre(), obrador.getNombre(), distanciaKm))
                .build();

        log.info("Distancia calculada: {} km", resultado.getDistanciaKm());
        return resultado;
    }

    /**
     * Fórmula de Haversine para calcular la distancia entre dos puntos en la esfera terrestre
     *
     * @param lat1 Latitud del punto 1 (en grados)
     * @param lon1 Longitud del punto 1 (en grados)
     * @param lat2 Latitud del punto 2 (en grados)
     * @param lon2 Longitud del punto 2 (en grados)
     * @return Distancia en kilómetros
     */
    private double calcularDistanciaHaversine(double lat1, double lon1, double lat2, double lon2) {
        // Convertir grados a radianes
        double lat1Rad = Math.toRadians(lat1);
        double lon1Rad = Math.toRadians(lon1);
        double lat2Rad = Math.toRadians(lat2);
        double lon2Rad = Math.toRadians(lon2);

        // Diferencias
        double deltaLat = lat2Rad - lat1Rad;
        double deltaLon = lon2Rad - lon1Rad;

        // Fórmula de Haversine
        double a = Math.sin(deltaLat / 2) * Math.sin(deltaLat / 2) +
                Math.cos(lat1Rad) * Math.cos(lat2Rad) *
                        Math.sin(deltaLon / 2) * Math.sin(deltaLon / 2);

        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

        // Distancia en kilómetros
        return RADIO_TIERRA_KM * c;
    }

    /**
     * Genera un mensaje descriptivo sobre la distancia
     */
    private String generarMensaje(String equipoNombre, String obradorNombre, double distanciaKm) {
        if (distanciaKm < 1) {
            return String.format("El equipo '%s' está a menos de 1 km del obrador '%s'",
                    equipoNombre, obradorNombre);
        } else if (distanciaKm < 10) {
            return String.format("El equipo '%s' está a %.2f km del obrador '%s' (distancia corta)",
                    equipoNombre, distanciaKm, obradorNombre);
        } else if (distanciaKm < 50) {
            return String.format("El equipo '%s' está a %.2f km del obrador '%s' (distancia media)",
                    equipoNombre, distanciaKm, obradorNombre);
        } else {
            return String.format("El equipo '%s' está a %.2f km del obrador '%s' (distancia larga)",
                    equipoNombre, distanciaKm, obradorNombre);
        }
    }
}